cmd/envtool/tests.go

```diff
@@ -363,21 +363,30 @@ func testsRun(ctx context.Context, index, total uint, run, skip string, args []s
 		return fmt.Errorf("--shard-index and --shard-total must be specified when --run is not")
 	}
 
-	all, err := listTestFuncsWithRegex("", run, skip)
+	tests, err := listTestFuncsWithRegex("", run, skip)
 	if err != nil {
 		return lazyerrors.Error(err)
 	}
 
-	if len(all) == 0 {
-		return fmt.Errorf("no tests to run")
-	}
-
-	shard, err := shardTestFuncs(index, total, all)
+	// Then, shard all the tests but only run the ones that match the regex and that should
+	// be run on the specific shard.
+	shard, skipShard, err := shardTestFuncs(index, total, tests)
 	if err != nil {
 		return lazyerrors.Error(err)
 	}
 
-	args = append(args, "-run="+buildGoTestRunRegex(shard))
+	args = append(args, "-run="+run)
+
+	if len(skipShard) > 0 {
+		if skip != "" {
+			skip += "|"
+		}
+		skip += "^(" + strings.Join(skipShard, "|") + ")$"
+	}
+
+	if skip != "" {
+		args = append(args, "-skip="+skip)
```

Comment:
I don't understand what we are doing there. #3969's title is "Process `--run` and `--skip` in the `envtool` itself". Somehow, we both process those flags ourselves, then ignore the result for the `--run` processing, but pass `-skip`.

What I think we should do is what the issue says. In more detail:
* get `--run` and `--skip` values;
* get a list of all tests;
* filter that list by applying `--run` and `--skip` values;
* run those tests by invoking `go test -run` – without `-skip` because we already process `--skip`'s regular expression.